src/main/java/uk/gov/ons/ctp/common/config/OverrideableYmlConfigReader.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    return cleanedEnvVariables;
+  }
+
+  private String cleanEnvironmentVariableName(String name) {
+    return name.replaceAll("(\\.|-|_)", "#").toLowerCase();
+  }
+
+  // Get Jackson to read YML file
+  private JsonNode readYmlFile(String resourcePath) throws CTPException {
+    JsonNode config;
+    try {
+      ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+      File resourceFile = new File(getClass().getClassLoader().getResource(resourcePath).getFile());
+      config = mapper.readTree(resourceFile);
+    } catch (IOException e) {
+      String errorMessage = "Failed to read contents of configuration file: " + resourcePath;
+      log.error(errorMessage);
+      throw new CTPException(Fault.SYSTEM_ERROR, e, errorMessage);
+    }
+    return config;
+  }
+
+  // Recursively search for properties to be updated
+  private void traverseAndUpdateProperties(JsonNode node, String pathToNode) throws CTPException {
+    Iterator<Entry<String, JsonNode>> iter = node.fields();
+    while (iter.hasNext()) {
+      Entry<String, JsonNode> entry = (Entry<String, JsonNode>) iter.next();
+      String key = entry.getKey();
+      JsonNode childNode = entry.getValue();
+
+      if (childNode.isArray()) {
+        // Ignore, as updates on properties not supported
+      } else if (childNode.isContainerNode()) {
+        traverseAndUpdateProperties(childNode, createPathToNode(pathToNode, key));
+      } else {
+        // Found a property
+        attemptNodeUpdate(node, pathToNode, key);
+      }
+    }
+    ;
+  }
+
+  private void attemptNodeUpdate(JsonNode parentNode, String pathToParent, String nodeName)
+      throws CTPException {
+    // Build full name of node, eg 'rabbit#host'
+    String rawPathToNode = createPathToNode(pathToParent, nodeName);
+    String pathToNode = cleanEnvironmentVariableName(rawPathToNode);
+
+    if (envVariables.containsKey(pathToNode)) {
+      ObjectNode objectNode = (ObjectNode) parentNode;
+      if (!objectNode.has(nodeName)) {
+        String errorMessage = "Error: Attempting to set value for unknown node: '" + nodeName + "'";
+        log.error(errorMessage);
+        throw new CTPException(Fault.SYSTEM_ERROR, errorMessage);
+      }
+
+      // Update value of node
+      String replacementValue = envVariables.get(pathToNode);
+      objectNode.put(nodeName, replacementValue);
+    }
+  }
+
+  private String createPathToNode(String parentNodePath, String childNodeName) {
+    if (parentNodePath.isEmpty()) {
+      return childNodeName;
+    } else {
+      return parentNodePath + "_" + childNodeName;
+    }
+  }
+
+  /**
+   * This method uses Jackson to convert the property data into a Java object.
+   *
+   * @param clazz is the class to convert the data into.
+   * @return Object containing the configuration date.
+   * @throws CTPException if the conversion failed.
+   */
+  public <T> T convertToObject(Class<T> clazz) throws CTPException {
```

Comment:
being picky but public methods at top of class between constructor and privates

Reply:
done